GH-41246: [Docs][C++][Python] Improve docs on column encryption for nested fields (#45411)
apache/arrow: docs/source/cpp/parquet.rst

* EncryptionWithFooterKey and EncryptionWithColumnKey modes.
* Encrypted Footer and Plaintext Footer modes.

Configuration
~~~~~~~~~~~~~

Parquet encryption uses a ``parquet::encryption::CryptoFactory`` that has access to a
Key Management System (KMS), which stores actual encryption keys, referenced by key ids.
The Parquet encryption configuration only uses key ids, no actual keys.

Parquet metadata encryption is configured via ``parquet::encryption::EncryptionConfiguration``:

.. literalinclude:: ../../../cpp/examples/arrow/parquet_column_encryption.cc
   :language: cpp
   :start-at: // Set write options with encryption configuration
   :end-before: encryption_config->column_keys
   :dedent: 2

> All columns are encrypted with the same key as the Parquet metadata when above
> ``encryption_config->uniform_encryption`` is set ``true``.
> 
> Individual columns are encrypted with individual keys as configured via ``encryption_config->column_keys``.
> This field expects a string of the format ``"columnKeyID:colName,colName;columnKeyID:colName..."``.

Reviewer: I would rephrase this more clearly:
```suggestion
If ``encryption_config->uniform_encryption`` is set to ``true``, then all columns are
encrypted with the same key as the Parquet metadata. Otherwise, individual
columns are encrypted with individual keys as configured via
``encryption_config->column_keys``. This field expects a string of the format
``"columnKeyID1:colName1,colName2;columnKeyID3:colName3..."``.
```

.. literalinclude:: ../../../cpp/examples/arrow/parquet_column_encryption.cc
   :language: cpp
   :start-at: // Set write options with encryption configuration
   :end-before: auto parquet_encryption_config
   :emphasize-lines: 4-5
   :dedent: 2

See the full `Parquet column encryption example <examples/parquet_column_encryption.html>`_.

.. note::

>    Encrypting columns that have nested fields (struct, map, or even list data types)
>    requires column keys for the inner fields, not the column itself.
>    Configuring a column key for the column itself causes this error (here column name is ``col``):

Reviewer: ```suggestion
   Encrypting columns that have nested fields (struct, map or list data types)
   requires column keys for the inner fields, not the nested column itself.
   Configuring a column key for the nested column causes
   this error (here the column name is ``col``):
```

Author: I find `nested column` ambiguous, it sounds like the inner column, the nested field.

What about `outer column` instead of `nested column`?

```suggestion
   Encrypting columns that have nested fields (struct, map, or even list data types)
   requires column keys for the inner fields, not the outer column itself.
   Configuring a column key for the outer column causes this error (here column name is ``col``):
```

Reviewer: You're right, that's better!

Reviewer: Also two wording nits:
- "struct, map, or even list data types" -> "struct, map, or list data types" (there's nothing special about lists)
- "here column name is" -> "here the column name is"

Author: Sorry, missed those of your changes when creating my suggestion. All incorporated now.

   .. code-block::

      OSError: Encrypted column col not in file schema

>    The key and value fields of a map column ``m`` has the names ``m.key_value.key``
>    and  ``m.key_value.value``, respectively. The inner field of a list column ``l``
>    has the name ``l.list.element``. An inner field ``f`` of a struct column ``s`` has
>    the name ``s.f``.

Reviewer: ```suggestion
   Conventionally, the key and value fields of a map column ``m`` have the names
   ``m.key_value.key`` and  ``m.key_value.value``, respectively. The inner field of a
   list column ``l`` has the name ``l.list.element``. An inner field ``f`` of a struct column ``s`` has
   the name ``s.f``.
```

* EncryptionWithFooterKey and EncryptionWithColumnKey modes.
* Encrypted Footer and Plaintext Footer modes.

Configuration
~~~~~~~~~~~~~

An example for writing a dataset using encrypted Parquet file format:

> .. code-block:: cpp

Reviewer: @jorisvandenbossche @AlenkaF @raulcd What is our preferred policy for code examples? Do we put them inline in the docs? Do we use separate files?

Reviewer: Update: it seems we use `literalinclude` directives from C++ example files that are compiled as part of CI runs. See for example https://github.com/apache/arrow/blob/main/docs/source/cpp/dataset.rst#reading-datasets

   // Directly assign shared_ptr objects to ParquetEncryptionConfig members
   parquet_encryption_config->crypto_factory = crypto_factory;
   parquet_encryption_config->kms_connection_config = kms_connection_config;
   parquet_encryption_config->encryption_config = std::move(encryption_config);

   auto file_format = std::make_shared<ParquetFileFormat>();
   auto parquet_file_write_options =
       checked_pointer_cast<ParquetFileWriteOptions>(file_format->DefaultWriteOptions());
   parquet_file_write_options->parquet_encryption_config =
       std::move(parquet_encryption_config);

   // Write dataset.
   arrow::Table table = ...;
   auto dataset = std::make_shared<InMemoryDataset>(table);
   EXPECT_OK_AND_ASSIGN(auto scanner_builder, dataset->NewScan());
   EXPECT_OK_AND_ASSIGN(auto scanner, scanner_builder->Finish());

   FileSystemDatasetWriteOptions write_options;
   write_options.file_write_options = parquet_file_write_options;
   write_options.base_dir = "example.parquet";
   ARROW_CHECK_OK(FileSystemDataset::Write(write_options, std::move(scanner)));

Column encryption is configured by setting ``encryption_config->column_keys`` to a string

> of the format ``"masterKeyID:colName,colName;masterKeyID:colName..."``.

Reviewer: ```suggestion
of the format ``"columnKeyID:colName,colName;columnKeyID:colName..."``.
```

Author: Took this from the source documentation:
https://github.com/apache/arrow/blob/240ebc98a623a78a058c110a38d410686141f487/cpp/src/parquet/encryption/crypto_factory.h#L47

I prefer `columnKeyId` over `masterKeyId`, as it is more descriptive. The term `masterKey` is used throughout the code, but it is not visible to the user, so this should be fine. Maybe the source documentation linked above should be updated as well.

Reviewer: Any improvements in the source documentation are welcome, if you have time and motivation for them :)

Encrypting columns that have nested fields (for instance struct, map, or even list data types)
require configuring column keys for the inner fields, not the column itself.
Configuring a column key for the column itself causes this error (here column name is ``col``):

.. code-block::

   OSError: Encrypted column col not in file schema

An example encryption configuration for columns with nested fields:

.. code-block:: cpp

   auto table_schema = schema({
     field("ListColumn", list(int32())),
     field("MapColumn", map(utf8(), int32())),
     field("StructColumn", struct_({field("f1", int32()), field("f2", utf8())})),
   });

>    encryption_config->column_keys = "column_key_name: "

Reviewer: ```suggestion
   encryption_config->column_keys = "column_key_id: "
```

                                    "ListColumn.list.element, "
                                    "MapColumn.key_value.key, MapColumn.key_value.value, "

>                                     "StructColumn.f1, StructColumn.f2"

Reviewer: Are the spaces embedded in the string actually supported? Also, it seems to lack a semicolon at the end of the line.

Author: Yes it supports extra whitespaces, it trims the strings. The final semicolon is not needed, it would introduce a next empty section:
https://github.com/apache/arrow/blob/240ebc98a623a78a058c110a38d410686141f487/cpp/src/parquet/encryption/crypto_factory.cc#L101-L167
